### PR TITLE
Updated RQ signature

### PR DESCRIPTION
**Pull request by nikita-malininn**
### Changes

- Updated `ReferenceQuantize.backward` signature to align with the CUDA/CPU extensions signature.

### Reason for changes

- Bugfix

### Related tickets

- 166245

### Tests

- Added `tests/torch/quantization/test_functions.py`

### CODE REVIEW

**Comment by ljaljushkin** [comment]
> There's no test to catch the bug. Please add it. E.g. you can mock call of `Quantize_backward` and check arguments type via `call_args`. This test would fail before this PR and pass with it.

`call_args` probably won't help with the test, but still there's no test checking that extension initialized by reference code is functional. 
Before your PR the extension wasn't functional in that case and nobody has noticed that for several years until now.

**Comment by nikita-malininn** [comment]
> > There's no test to catch the bug. Please add it. E.g. you can mock call of `Quantize_backward` and check arguments type via `call_args`. This test would fail before this PR and pass with it.
> 
> `call_args` probably won't help with the test, but still there's no test checking that extension initialized by reference code is functional. Before your PR the extension wasn't functional in that case and nobody has noticed that for several years until now, because there was no test.

The test was added.

File: nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp [hunk 1]
```diff
@@ -50,16 +49,12 @@ std::vector<at::Tensor> q_cpu_backward(
...
-        grad_input_low = grad_input_low.masked_fill_(at::__xor__(all_ones, outside_mask), 0);
+    auto grad_input_low = grad_output.clone();
+    auto all_ones = torch::ones_like(outside_mask);
+    grad_input_low = grad_input_low.masked_fill_(at::__xor__(all_ones, outside_mask), 0);
 
-        sum_like(grad_input_low, input_low);
-        return {grad_input, grad_input_low, grad_input_range};
-    }
-    auto dummy_variable = torch::autograd::make_variable(at::empty(input_low.sizes()), true);
-    return {grad_input, dummy_variable, grad_input_range};
```
[the remark this answers is not in the document]
Comment (nikita-malininn) [reply nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp:62]: Even CPU and CUDA signatures are different. 
[nncf/torch/extensions/src/quantization/cuda/functions_cuda.cpp#L21-L28](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/extensions/src/quantization/cuda/functions_cuda.cpp#L21-L28) - **is_symmetric** variable is not presented for CUDA.
[nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp#L91-L99](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp#L91-L99) - **is_symmetric** variable is presented for CPU.
I doubt that **grad_input_low** is not presented in [CUDA kernels](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/extensions/src/quantization/cuda/functions_cuda_impl.cu) -, but based on the absence of **is_symmetric** variable, we may assume that the calculation exists here.

Thus, there are a few questions: why are calculations on CUDA not optimised for the symmetric case? Which scenario is correct then, since there are 3 options: calculate the gradient for **input_low** as common, do not calculate it, calculate based on the flag.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp:62]: No idea, why symmetric case wasn't optimized on `CUDA`. Better to ask @vshampor and @alexsu52.
However, this value is not utilized and can be any value that does not cause compilation errors: https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/quantize_functions.py#L65

Current code for the reference doesn't look optimal, because of some unnecessary manipulations with unused gradient. But you can double-check the performance by profiling. I'd return None or dummy tensor as before.

File: tests/torch/quantization/test_functions.py [hunk 2]
```diff
@@ -669,3 +672,99 @@ def test_get_scale_zp_from_input_low_input_high(
...
+        desc.level_high,
+    ]
+
+    ext_fwd_output = QuantizedFunctionsCPU.get("Quantize_forward")(*fwd_args)
+    ref_fwd_output = ReferenceQuantizedFunctions.Quantize_forward(*fwd_args)
+
+    assert torch.allclose(ext_fwd_output, ref_fwd_output)
+
+    bwd_grad_input, bwd_grad_low, bwd_grad_range = QuantizedFunctionsCPU.get("Quantize_backward")(*bwd_args)
+    ref_grad_input, ref_grad_low, ref_grad_range = ReferenceQuantizedFunctions.Quantize_backward(*bwd_args)
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:722]: Again, gradient for input low is not used anywhere as intended. IMHO, nothing wrong here and it's not an error.
It's coming from a requirement of `torch.autograd` to have the same number of outputs on backward as the number of inputs on forward. But symmetric case has only one learnable parameter - scale, which is represented by `input_range`, that's why only one gradient is used.

My ask was about testing the functionality of extensions with reference implementation. 
The gradient for input low doesn't influence this functionality. Hope, it makes things clear.

[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:722]: As you wish.

File: nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikita-malininn) [reply nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp:62]: As you wish.

File: tests/torch/quantization/test_functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:737]: However, currently, the method signatures are different between CUDA and CPU extensions.

Furthermore, the **grad_range** is not the same between CUDA and the reference execution. We can continue to sweep the issues under the rug and only check **grad_input**, if you want?

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:737]: Please stop misinterpreting my words. 
I didn't say that we should check only `grad_input` and say nothing about different signatures. 

My point was only about gradient for `input_low`. 
Feel free to align signatures, fix and check gradient for `input_range` - no objection for that.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:737]: Just to summarize and give more insights: 
1) alignment of gradient for `input_low` is optional, since it's not used anywhere. 
2) alignment of signature (Extra `is_asymmetric` argument), IMHO, is not critical, but no objections from my side to make it unified. 
3) failed test for CUDA should be investigated and fixed. If it's just different order of operations, some tolerance value can be used.

**Review by ljaljushkin — DISMISSED** [review]

File: tests/torch/quantization/test_functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:737]: What about the alignment for **grad_low** and **grad_range** output formats? Their formats differ for extensions and reference.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:737]: Yes, I meant that it's not used anywhere for symmetric case:
https://github.com/openvinotoolkit/nncf/pull/3441#discussion_r2049118691
For sure, in asymmetric case it makes sense to check all gradients: `grad_input`, `grad_input_low`, `grad_input_range`. 

> What about the alignment for grad_low and grad_range output formats?
What exactly is different and in which cases? Do you only mean the symmetric case and the dummy gradient for `input_low`, or are there other differences?

[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:737]: image
As you can see, the output formats for `bwd_grad_low` and `ref_grad_low`. 
The same situation is for `grad_input_range`. You can even observe that in failed tests:
image

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:737]: Yes, it looks as a bug, shape should be the same.

[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:737]: The main question which was asked multiple times: which behaviour is correct here - CPU ext., CUDA ext. or Reference?
Which shape and output value is correct?

Without this knowledge, it is not really possible to align anything with each other.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:737]: By definition, the gradient's shape matches the parameter's shape. 
Regarding output value: reference is the reference :)

File: tests/torch/quantization/test_functions.py [hunk 3]
```diff
@@ -669,3 +672,128 @@ def test_get_scale_zp_from_input_low_input_high(
...
+
+    assert torch.allclose(bwd_grad_input, ref_grad_input)
+    if desc.is_asymmetric:
+        assert torch.allclose(bwd_grad_low, ref_grad_low)
+    assert torch.allclose(bwd_grad_range, ref_grad_range)
+
+
+@pytest.mark.cuda
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="Skipping for CPU-only setups")
```
Comment (ljaljushkin) [inline tests/torch/quantization/test_functions.py:746]: no need in that. it's handled by marker

File: tests/torch/quantization/test_functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:737]: > Regarding output value: reference is the reference

I don't want to be misunderstood again, but just curious: I've changed CPU ext. regarding the Reference, and you said to roll back these changes. Now, your statement that extensions should be the same as the Reference. Are you sure?

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py:737]: > you said to roll back these changes
I didn't. 

Feel free to call me if you have questions. I doubt that this discussion is constructive.

File: tests/torch/quantization/test_functions.py (hunk 3 above)
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py]: Done.

File: tests/torch/quantization/test_functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py:737]: Tests passed with your fix. Thanks.

File: tests/torch/quantization/test_functions.py [hunk 4]
```diff
@@ -669,3 +672,128 @@ def test_get_scale_zp_from_input_low_input_high(
     )
     assert zero_point == ref_zero_point, f"{zero_point} != {ref_zero_point}"
     assert np.isclose(scale, ref_scale), f"{scale:.10f} != {ref_scale}"
+
+
+class CompatibilityTestDesc:
+    def __init__(self, levels, level_low, level_high, is_asymmetric=False):
+        self.input_ = torch.tensor([[-0.5, 0.5]])
+        self.input_low = torch.tensor([-0.5])
+        self.input_range = torch.tensor([1.0])
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/torch/quantization/test_functions.py]: Do you have understanding why it didn't work without it? Was it valid behavior? Is some fix required in `sum_like`?

[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py]: Not yet. I'll do my best to provide more details soon.

[the remark this answers is not in the document]
Comment (nikita-malininn) [reply tests/torch/quantization/test_functions.py]: I've checked the test outputs regarding the issue.
I think that `sum_like` operates correctly, and the reason for failures was the incorrect `input_low` and `input_range` shapes.
For example, if the `input_` shape `[1, 2]`, then the `low` and `range` shapes should be `[1, 1]` each, but not `[1]`.

**Review by alexsu52 — APPROVED** [review]
LGTM
